### PR TITLE
Maintenance stats for remote projects

### PR DESCRIPTION
**Pull request by bgeuken**

### CODE REVIEW

**Comment by bgeuken** [comment]
@adrianschroeter @evanrolfe Please review

**Comment by bgeuken** [comment]
btw I already tested with osc if it properly handles a redirect of the stats request to a remote instance (build-test). It works.

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb [hunk 1]
```diff
@@ -1,12 +1,24 @@
 module Statistics
   class MaintenanceStatisticsController < ApplicationController
-    skip_before_action :extract_user
+    skip_before_action :extract_user, :require_login
```
Comment (hennevogel) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: shouldn't this be in `/public` then?

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
     def index
       @project = Project.get_by_name(params[:project])
-      @maintenance_statistics = MaintenanceStatisticDecorator.wrap(
-        MaintenanceStatistic.find_by_project(@project)
-      )
+      if @project.is_a?(String)
+        remote_instance, remote_project = Project.find_remote_project(@project)
+        # Port get's explicitly set to properly redirect from dev machines (port 3000)
+        # to instances running in production mode.
+        redirect_to maintenance_statistics_url(
```
Comment (hennevogel) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: Hm this is not what was requested I fear. "Maintenance statistics of remote projects can be viewed on local instances". Redirecting is something else. Why don't you fetch the data from remote and render it?

File: src/api/spec/controllers/statistics/maintenance_statistics_controller_spec.rb [hunk 3]
```diff
@@ -37,6 +37,20 @@
           assert_response :not_found
         end
       end
+
+      context 'with a remote project' do
+        let(:remote) { create(:remote_project)}
```
Comment (Ana06) [inline src/api/spec/controllers/statistics/maintenance_statistics_controller_spec.rb]: space missed:
`let(:remote) { create(:remote_project) }`

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb (hunk 2 above)
Comment (bgeuken) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: Let's see what @adrianschroeter says. Though changing this shouldn't be a big deal.

Comment (bgeuken) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: > hennevogel 2 hours ago
> Why don't you fetch the data from remote and render it?

I don't see the point of doing additional work on the local instance. It's not like the local host would have to process the data in any way. We pass them through anyway, so why not redirecting the request?

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: I can move the public routing there, if you want. Just note that we already route calls to public/* to the respective controllers.

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb (hunk 2 above)
Comment (hennevogel) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: @adrianschroeter do you have an opinion?

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb (hunk 1 above)
Comment (hennevogel) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: Sure I just mean the routing.

Comment (bgeuken) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: That's already handled. Though I kept the existing route as well, to stay compatible

File: src/api/app/controllers/statistics/maintenance_statistics_controller.rb (hunk 2 above)
Comment (bgeuken) [inline src/api/app/controllers/statistics/maintenance_statistics_controller.rb]: @adrianschroeter what do you say?

**Comment by adrianschroeter** [comment]
I am not a in favor that we handle the remote case in this particular case different then in other places, meaning sending a redirect back and rely on the client implementation. This is at least an inconsistency.

So, if it is not too much work, I would prefer if we proxy the request here as well.

We can plan for 3.0 an overall different behaviour, if we see an advantage in doing redirect in such cases.

**Comment by bgeuken** [comment]
@adrianschroeter Updated. Though I have to say that I liked the version with the simple redirect better. just for the record^^

**Comment by adrianschroeter** [comment]
@bgeuken understood, so we should collect this as discussion point for 3.0 at least ...

**Comment by hennevogel** [comment]
@Ana06?

**Pull request merged**